### PR TITLE
feat(seer): Score non-top predictions properly, score team/shared-team differently

### PR DESCRIPTION
**Pull request by hobzcalvin**
We sometimes correctly choose a lower-ranked mapped user as our prediction over a higher-ranked unmapped user, but still score that as a `MISS`. So, in addition to the full prediction ranks, store `selected_assignee_user_id` and score against that instead. Note that if we pick the user in the #2 slot, `hit_rank` will still be 2.

Also, branch the `TEAM` result (which is almost certainly a majority of predicted-user-in-assigned-team) off into the less-frequent/higher-signal variant `SHARED_TEAM` (the actual assigned user shares a team with the predicted user).

Relates to ISWF-2943

### CODE REVIEW

File: src/sentry/seer/smart_assignment/completion.py [hunk 1]
```diff
@@ -27,17 +27,29 @@ def process_smart_assignment_completion(group: Group, activity: Activity) -> Non
...
+    # down the list of suggestions if necessary.
+    top_user_id, used_rank = next(
+        (
+            (user_id, rank)
+            for rank, user_id in enumerate(predicted_assignee_user_ids)
+            if user_id is not None
+        ),
+        (None, None),
+    )
+
```
Comment (hobzcalvin) [inline src/sentry/seer/smart_assignment/completion.py:40]: Just moved this from inside `_apply_prediction()` below

File: src/sentry/seer/smart_assignment/scoring.py [hunk 2]
```diff
@@ -171,9 +182,14 @@ def _apply(run_id: int, updates: RunUpdates) -> bool:
...
         # Score the prediction against the ground truth if it's already landed.
+        predicted_user_ids = extras.get("predicted_assignee_user_ids") or []
         result, hit_rank = _score(
             run.run.organization_id,
-            predicted_user_ids=extras.get("predicted_assignee_user_ids") or [],
+            predicted_user_ids=predicted_user_ids,
+            selected_user_id=extras.get(
+                "selected_assignee_user_id",
+                predicted_user_ids[0] if predicted_user_ids else None,
+            ),
```
Comment (saponifi3d) [inline src/sentry/seer/smart_assignment/scoring.py:192]: the def for `predicted_user_ids` is `list[int | None]` so we might want to handle if an element in the list is also none, rather than selecting the first id.

Here's a quick change that we could do to select the first defined id.
```python
selected_user_id=extras.get(
    "selected_assignee_user_id",
    next((user_id for user_id in predicted_user_ids if user_id is not None), None),
)
```

if you want to introduce a unit test for this, you can have the `predicted_user_ids = [None, 1234]`.

File: tests/sentry/seer/smart_assignment/test_scoring.py [hunk 3]
```diff
@@ -118,27 +159,15 @@ def test_no_hit_rank_when_actual_absent_from_candidates(self, mock_metrics: Magi
         alice = self.create_user()
         bob = self.create_user()
         run = self._run(actual_assignee_user_id=bob.id)
-        record_prediction(run, [alice.id])
+        record_prediction(run, [alice.id], alice.id)
```
Comment (saponifi3d) [inline tests/sentry/seer/smart_assignment/test_scoring.py:162]: The missing test case here would be: `[None, alice.id]` -- this would currently return `None`, where it probably should return alice.id

**Review by saponifi3d — APPROVED** [review]
lgtm - i'd just do the fix with `next` on the predicted_user_ids

File: src/sentry/seer/smart_assignment/scoring.py (hunk 2 above)
Comment (hobzcalvin) [reply src/sentry/seer/smart_assignment/scoring.py:192]: I'll use the `next()` thing, but that's only a fallback for the runs we score after this change is deployed: `seleted_assignee_user_id` will always be set from now on, and it _can_ be `None` if we didn't select anyone. Currently we score against the top slot, even if it's `None` and we actually fell back to a lower-ranked choice. It _is_ more correct to score against that lower-ranked choice, so your change makes sense.
